src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -33,6 +33,11 @@ namespace Microsoft.CodeAnalysis
     public abstract partial class Compilation
     {
         /// <summary>
+        /// For use only in <see cref="Microsoft.CodeAnalysis.Collections.CompilationStuff"/>.
+        /// </summary>
+        internal ConcurrentDictionary<object, object> _compilationStuff = new ConcurrentDictionary<object, object>();
```

Comment:
Actually I think @pharring has a point here. I believe, IDE often mutates compilation by using With... methods, adds syntax trees, changes options, adds references, etc. That is why we are trying to reuse as much computed data as possible when we implement these methods. Allocating a new dictionary per each call like this is probably an overkill.